### PR TITLE
Fix use of thread-specific distributions in random parameters

### LINKED ISSUE #2119: Potjans_2014 example in NEST 3.0 gives different output under different MPI-process/thread splits with fixed VP number
**Describe the bug**
Simulation of the Potjans_2014 microcircuit from pynest/example/Potjans_2014 using 4 MPI-processes and 1 thread per MPI process (total of 4 VP) and 1 MPI-process with 4 threads per MPI process (total of 4 VP) gave different results. Other parameters were kept constant, including the seed witch was the default 55.

**To Reproduce**
1. (Minimal) reproducing example
1.1. 4 MPI processes with 1 thread per MPI process:
- Copy the Potjans_2014 code located in pynest/example/Potjans_2014
- simulate using 4 MPI processes (nothing was changed)
 -- mpirun -np 4 python3 ./run_microcircuit.py
- Output:
>$ mpirun -np 4 python3 ./run_microcircuit.py 
> 
>              -- N E S T --
>  Copyright (C) 2004 The NEST Initiative
>
> Version: nest-3.0
> ...
>Neuron numbers are scaled by a factor of 0.100.
>Indegrees are scaled by a factor of 0.100.
>  Weights and DC input are adjusted to compensate.
>
>RNG seed: 55 
> ...
>  Total number of virtual processes: 4
> ...
>Creating neuronal populations.
>Creating recording devices.
>  Creating spike recorders.
>Creating Poisson generators for background input.
>Connecting neuronal populations recurrently.
>Connecting Poisson generators for background input.
>
>Aug 14 15:56:37 NodeManager::prepare_nodes [Info]: 
>    Preparing 1945 nodes for simulation.
> ...
>Simulating 500.0 ms.
> ...
>    Simulation time (ms): 500
>    Number of OpenMP threads: 1
>    Number of MPI processes: 4
> ...
>Simulating 1000.0 ms.
> ...
>    Simulation time (ms): 1000
>    Number of OpenMP threads: 1
>    Number of MPI processes: 4
> ...
>    Simulation finished.
> ...
>Interval to compute firing rates: [ 500. 1500.] ms
>Mean rates: [0.419 1.997 3.979 4.909 6.449 7.66  0.775 6.953] spikes/s
>Standard deviation of rates: [0.733 1.688 2.75  3.306 4.056 4.412 1.082 3.679] spikes/s
> ...

1.2. 1 MPI processes with 4 thread per MPI process:
- Copy the Potjans_2014 code located in pynest/example/Potjans_2014
- Edit line 49 in sim_params.py to `'local_num_threads': 4,` 
- simulate using 1 MPI processes (everything else was kept unchanged)
 -- mpirun -np 1 python3 ./run_microcircuit.py
- Output:
>$ mpirun -np 1 python3 ./run_microcircuit.py 
>
>              -- N E S T --
>  Copyright (C) 2004 The NEST Initiative
>
> Version: nest-3.0
> ...
>Neuron numbers are scaled by a factor of 0.100.
>Indegrees are scaled by a factor of 0.100.
>  Weights and DC input are adjusted to compensate.
>
>RNG seed: 55 
>  Total number of virtual processes: 4
> ...
>Creating neuronal populations.
>Creating recording devices.
 > Creating spike recorders.
>Creating Poisson generators for background input.
><
>Connecting neuronal populations recurrently.
>Connecting Poisson generators for background input.
>
>Aug 14 15:56:06 NodeManager::prepare_nodes [Info]: 
>    Preparing 7781 nodes for simulation.
>Simulating 500.0 ms.
>
>Aug 14 15:56:07 NodeManager::prepare_nodes [Info]: 
>    Preparing 7781 nodes for simulation.
>
>Aug 14 15:56:07 SimulationManager::start_updating_ [Info]: 
>    Number of local nodes: 7781
>    Simulation time (ms): 500
>    Number of OpenMP threads: 4
>    Number of MPI processes: 1
> ...
>Simulating 1000.0 ms.
>
>Aug 14 15:56:14 NodeManager::prepare_nodes [Info]: 
>    Preparing 7781 nodes for simulation.
>
>Aug 14 15:56:14 SimulationManager::start_updating_ [Info]: 
>    Number of local nodes: 7781
>    Simulation time (ms): 1000
>    Number of OpenMP threads: 4
>    Number of MPI processes: 1
> ...
>    Simulation finished.
>Interval to plot spikes: [600. 800.] ms
>Interval to compute firing rates: [ 500. 1500.] ms
>Mean rates: [0.414 2.019 4.064 4.956 6.534 7.679 0.759 7.01 ] spikes/s
>Standard deviation of rates: [0.745 1.73  2.843 3.452 4.033 4.02  1.1   4.073] spikes/s
> ...

- 
2. Used parameters
Only local_num_threads was changed

3. Command to run
3.1 `$ mpirun -np 4 python3 ./run_microcircuit.py`
3.2 `$ mpirun -np 1 python3 ./run_microcircuit.py`
local_num_threads was changed from 1 to 4 in line 49 of sim_params.py
4. How to see error
Compare mean rates and Standard deviation of rates reported by the above command
**Expected behavior**
Since the total number of virtual processes (VP) are the same, and all other parameters, including seed, are unchanged, both runs should result in equal outputs, including spike times which also differed.
**Screenshots**

**Desktop/Environment (please complete the following information):**
 - OS: Fedora 34
 - Shell: bash
 - Python-Version: Python 3.9
 - NEST-Version: nest-3.0
 - Installation: 
>--------------------------------------------------------------------------------
>NEST Configuration Summary
>--------------------------------------------------------------------------------
>
>Build type          : 
>Target System       : Linux
>Cross Compiling     : FALSE
>C compiler          : GNU 11.1.1 (/usr/bin/cc)
>C compiler flags    :  -O2 -Wall -fopenmp   -pthread -fdiagnostics-color=auto
>C++ compiler        : GNU 11.1.1 (/usr/bin/c++)
>C++ compiler flags  :  -std=c++11 -O2 -Wall -fopenmp -pthread -fdiagnostics-color=auto
>Build dynamic       : ON
>
>Built-in modules    : models
>User modules        : None
>Python bindings     : Yes (Python 3.9.6: /usr/bin/python3.9)
>       Includes     : /usr/include/python3.9
>       Libraries    : /usr/lib64/libpython3.9.so
>
>Cython bindings     : Yes (Cython 0.29.21: /usr/bin/cython)
>MPI4PY header	    : Yes (/usr/lib64/python3.9/site-packages/openmpi/mpi4py/include)
>Use threading       : Yes (OpenMP: -fopenmp)
>Use GSL             : Yes (GSL 2.6)
>    Includes        : /usr/include
>    Libraries       : /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so
>
>Use Readline        : Yes (GNU Readline 8.1)
>    Includes        : /usr/include
>    Libraries       : /lib64/libreadline.so;/lib64/libncurses.so
>
>Use libltdl         : Yes (LTDL )
>    Includes        : /usr/include
>    Libraries       : /lib64/libltdl.so
>
>Use doxygen         : Yes (/usr/bin/doxygen)
>                    : target `doc` available
>    `dot` available : Yes (/usr/bin/dot)
>                    : target `fulldoc` available
>Use MPI             : Yes (MPI: /usr/lib64/openmpi/bin/mpicxx)
>    FLAGS           : -pthread
>    Includes        : /usr/include/openmpi-x86_64
>    Link Flags      : -Wl,-rpath -Wl,/usr/lib64/openmpi/lib -Wl,--enable-new-dtags -pthread
>    Libraries       : /usr/lib64/openmpi/lib/libmpi_cxx.so;/usr/lib64/openmpi/lib/libmpi.so
>
>Detailed timers     : No
>
>Use MUSIC           : No
>Use libneurosim     : No
>Use Boost           : Yes (Boost 1.75.0)
>    Includes        : /usr/include
>    Libraries       : 
>
>Use recording backend Arbor   : No
>Use SIONlib         : No
>
>--------------------------------------------------------------------------------

**Additional context**
Same results were obtained in current master:
1. 4 MPI processes with 1 thread per MPI
>$ mpirun -np 4 python3 ./run_microcircuit.py 
> ...
> Version: master@c4a0daab5
> ...
>Interval to compute firing rates: [ 500. 1500.] ms
>Mean rates: [0.419 1.997 3.979 4.909 6.449 7.66  0.775 6.953] spikes/s
>Standard deviation of rates: [0.733 1.688 2.75  3.306 4.056 4.412 1.082 3.679] spikes/s

2. 1 MPI process with 4 threads per MPI
>$ mpirun -np 1 python3 ./run_microcircuit.py 
> ...
> Version: master@c4a0daab5
> ...
>Interval to compute firing rates: [ 500. 1500.] ms
>Mean rates: [0.414 2.019 4.064 4.956 6.534 7.679 0.759 7.01 ] spikes/s
>Standard deviation of rates: [0.745 1.73  2.843 3.452 4.033 4.02  1.1   4.073] spikes/s

### PR DESCRIPTION
**Pull request by hakonsbm**
Fixes #2119 by correcting the use of thread-specific distributions in the normal and log-normal random parameters. 

Also adds a test that checks that values from random distribution parameters are invariant with different number of MPI processes and threads, and fixed number of virtual processes.

### CODE REVIEW

File: nestkernel/parameter.cpp
Comment (heplesser) [inline nestkernel/parameter.cpp:106]: This makes sense, but I am a bit worried about performance and code duplication. We do exactly the same operations here:

https://github.com/nest/nest-simulator/blob/062e2723e1df19433e4c0131f1aa36a428067156/libnestutil/dict_util.h#L56-L57

right before calling this method. Both steps involve modulus or division, which are expensive. Have you checked the effect on perfomance? How about passing the `tid` from the calling function? Or would that cause problems elsewhere? Also, in `dict_util.h`, we could add `const` qualifiers which may allow more aggressive optimisation of chained modulus and division.

**Review by heplesser — COMMENTED** [review]
@hakonsbm Thanks for this PR. It seems some tests are still failing and I have one concern, see below.

**Comment by hakonsbm** [comment]
@heplesser I've updated the implementation, so the SLI test is no longer failing. 

I've also done some simple benchmarks with `timeit`:
```python
n = nest.Create('iaf_psc_alpha', 100000)
p = nest.random.normal()
%timeit n.V_m = p
```
Which results in
| master | this PR |
|-----------|----------|
| 759 ms ± 28.5 ms | 757 ms ± 26.1 ms |

So it seems like the compiler does some optimisations on the division, and performance isn't an issue here.

Although it seems like all tests are passing now, this is not the case. One PyNEST test fails with a segfault (hinted at in [this line](https://github.com/nest/nest-simulator/runs/3379615413#step:16:33746)), but is not reported in the installcheck results. In fact, no PyNEST tests, except for the new MPI ones, are reported there, so it may be that failing PyNEST tests are going unnoticed. I can't see the same problem in other recent PRs, but I don't think I've touched any files that should change this behaviour. @jougs Since you refactored the testsuite recently, could you take a look?

**Comment by jougs** [comment]
Even though I don't think that this is related to my recent refactoring of the PyNEST testsuite, I can indeed reproduce the segfault locally by running `python install/share/nest/testsuite/pytests/test_nodeParametrization.py`. Here's a backtrace obtained from `gdb`:

```gdb
Test positions specified by parameter ... 
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
nest::VPManager::node_id_to_vp (node_id=<optimized out>, this=<optimized out>)
    at /home/jochen/tmp/nest-simulator/nestkernel/random_generators.h:296
296	  inline result_type operator()( RngPtr g )
(gdb) bt
#0  nest::VPManager::node_id_to_vp (node_id=<optimized out>, this=<optimized out>)
    at /home/jochen/tmp/nest-simulator/nestkernel/random_generators.h:296
#1  nest::NormalParameter::value (this=0x1e363c0, rng=0x1f6fef0, node=0x0)
    at /home/jochen/tmp/nest-simulator/nestkernel/parameter.cpp:105
#2  0x00007ffff4e425c0 in nest::DimensionParameter::get_values (this=this@entry=0x1c16fc0, 
    rng=rng@entry=0x1f6fef0) at /home/jochen/tmp/nest-simulator/nestkernel/parameter.h:1234
#3  0x00007ffff4e4942b in nest::FreeLayer<2>::set_status (this=0x1b03350, d=...)
    at /home/jochen/tmp/nest-simulator/nestkernel/free_layer.h:178
#4  0x00007ffff4e4076e in nest::AbstractLayer::create_layer (layer_dict=...)
    at /home/jochen/tmp/nest-simulator/nestkernel/layer.cpp:168
#5  0x00007ffff4e6476c in nest::create_layer (layer_dict=...)
    at /home/jochen/tmp/nest-simulator/nestkernel/spatial.cpp:118
#6  0x00007ffff4d93fd7 in nest::NestModule::CreateLayer_D_DFunction::execute (this=<optimized out>, 
    i=0xe363e0) at /home/jochen/tmp/nest-simulator/nestkernel/nestmodule.cpp:2145
#7  0x00007ffff4bbbe02 in SLIInterpreter::execute_ (this=0xe363e0, exitlevel=0)
    at /home/jochen/tmp/nest-simulator/sli/interpret.cc:1339
#8  0x00007ffff54e0f7d in __pyx_pf_12pynestkernel_10NESTEngine_8run (__pyx_v_cmd=<optimized out>, 
    __pyx_v_self=0x7ffff7370f70) at /home/jochen/tmp/nest-simulator.build/pynest/pynestkernel.cxx:5689
#9  __pyx_pw_12pynestkernel_10NESTEngine_9run (__pyx_v_self=0x7ffff7370f70, 
    __pyx_v_cmd=<optimized out>) at /home/jochen/tmp/nest-simulator.build/pynest/pynestkernel.cxx:5582
```

Could the `node=0x0` in frame #⁠1 be the culprit here?

**Comment by hakonsbm** [comment]
@jougs Sorry, I meant the problem with PyNEST tests not showing up in the testsuite results and failing tests not causing the testsuite to fail. The segfault is because of this PR and I'm already working on fixing it.

**Comment by niltonlk** [comment]
I confirm that @hakonsbm branch solves #2119 

1. fix_normal_random branch 4 MPI w/ 1 OpenMP:
```
$ mpirun -np 4 python3 ./run_microcircuit.py

              -- N E S T --
  Copyright (C) 2004 The NEST Initiative

 Version: fix_normal_random@3dc3d38bb

Interval to compute firing rates: [ 500. 1500.] ms
Mean rates: [0.419 1.997 3.979 4.909 6.449 7.66  0.775 6.953] spikes/s
Standard deviation of rates: [0.733 1.688 2.75  3.306 4.056 4.412 1.082 3.679] spikes/s
```

2. fix_normal_random branch 1 MPI w/ 4 OpenMP:
```
$ mpirun -np 1 python3 ./run_microcircuit.py 

              -- N E S T --
  Copyright (C) 2004 The NEST Initiative

 Version: fix_normal_random@3dc3d38bb

Interval to compute firing rates: [ 500. 1500.] ms
Mean rates: [0.419 1.997 3.979 4.909 6.449 7.66  0.775 6.953] spikes/s
Standard deviation of rates: [0.733 1.688 2.75  3.306 4.056 4.412 1.082 3.679] spikes/s
```

Performance with the 0.1 scaled network seems to be small:

1. 4 MPI w/ 1 OpenMP:

|                                     | NEST 3.0 | fix_normal_random |
|------------------------------|---------------|----------------------------|
|  Total time:                  | 6.579 s     | 6.950 s                    | 
|  Time to initialize:       | 0.064 s     | 0.036 s                     |
|  Time to create:          | 0.074 s     | 0.071 s                     |
|  Time to connect:       | 0.484 s     | 0.498 s                     |
|  Time to presimulate: | 1.559 s     | 1.805 s                     |
|  Time to simulate:      | 3.344 s     | 3.482 s                     |
|  Time to evaluate:      | 1.053 s     | 1.058 s                     |

2. 1 MPI w/ 4 OpenMP:

|                                     | NEST 3.0 | fix_normal_random |
|------------------------------|---------------|----------------------------|
|  Total time:                  | 19.195 s   | 19.542 s                   | 
|  Time to initialize:       | 0.004 s     | 0.004 s                     |
|  Time to create:          | 0.138 s     | 0.138 s                     |
|  Time to connect:       | 1.526 s     | 1.543 s                     |
|  Time to presimulate: | 5.615 s     | 5.683 s                     |
|  Time to simulate:      | 10.965 s   | 11.230 s                    |
|  Time to evaluate:      | 0.946 s     | 0.945 s                     |

Now, why is the OpenMP parallelization taking about 3 times longer than the MPI parallelization? More surprisingly, OpenMP parallelization was slower than no parallelization at all?:

3. 1 MPI w/ 1 OpenMP:

|                                     | NEST 3.0 | fix_normal_random |
|------------------------------|---------------|----------------------------|
|  Total time:                  | 14.133 s   | 14.084 s                   | 
|  Time to initialize:       | 0.002 s     | 0.002 s                     |
|  Time to create:          | 0.129 s     | 0.123 s                     |
|  Time to connect:       | 1.366 s     | 1.382 s                     |
|  Time to presimulate: | 4.028 s     | 3.913 s                     |
|  Time to simulate:      | 7.648 s     | 7.729 s                     |
|  Time to evaluate:      | 0.940 s     | 0.935 s                     |

**Comment by jougs** [comment]
Ah, OK!

I did not change anything about how errors are caught by the testsuite. IIRC, this was last touched by @heplesser in #1587. However, I am pretty sure we have no mechanism for catching segfaults.

**Comment by heplesser** [comment]
@niltonlk Thank you for your benchmarks. I cannot reproduce your slow threaded performance. I tested with with current master (062e272) on my MacBook Pro with quad-core i5, compiled with Apple Clang 12.0.5 under OSX 11.5.1. I ran the PD benchmark out of the box. The only change I made was to set `local_num_threads` to 4 for the 1 MPI-4 Thread test.

I obtained the following results from three separate runs (Total time):

- 1 MPI, 1 Thread: 15.074 s, 15.572 s, 15.002 s
- 4 MPI, 1 Thread: 5.936, s 5.722, s 5.718 s
- 1 MPI, 4 Threads: 5.785 s, 5.645 s, 6.225 s

The times for the first two cases are similar to your results, but in my case threaded performance is on par with MPI performance. Could you check your configuration again to see if there may have been something preventing successful threading?

**Comment by heplesser** [comment]
@jougs @hakonsbm We should definitely catch segfaulting tests ;). I will take a look at this. The problem likely occurs because we are not running the mpitests directly from nosetest but through indirections ...

**Comment by heplesser** [comment]
@hakonsbm @jougs My analysis so far
- If a segfault occurs in any process run from `nosetests`, no XML output is written at the end, `nosetests` just crashes (note that in the log @hakonsbm linked to, only a single pytest is reported, the one from the MPI Python tests).
- In this case, `nosetests` returns with 139, while it returns with 1 when tests fail.
- We could use the latter if we could catch that return code [in do_tests.sh](https://github.com/nest/nest-simulator/blob/8c16625f39737318fceb97989e37fe6ab2b01342/testsuite/do_tests.sh#L463-L465), but my shell skills aren't up to getting the right return code from a pipeline ...
- Another problem is that the new test by @hakonsbm is not run at all, because it is not yet entered [in the list of Python MPI-tests](https://github.com/hakonsbm/nest-simulator/blob/3dc3d38bb459895b49e44eaf258263d99a100680/testsuite/pytests/test_mpitests.py#L44-L53). I tried to add it but then got some very confusing error output.

**Comment by niltonlk** [comment]
@heplesser I have done the tests in another environment, but the difference still persists The result presented above is for a Laptop with 4 physical cores running Fedora 34 with GCC 11.2.1, OpenMPI 4.1.0, Python 3.9.6, Boost 1.75 and gsl 2.6.

The second system is a Server with 2 CPU with a total of 20 physical cores running CentOS 7.9.2009 with GCC 8.3.0, OpenMPI 3.1.4, Python 3.8.10, Boost 1.71 and gsl 2.6. 

Below is the result for the second system only.

- 1 MPI, 1 Thread: 14.882 s, 13.135 s, 13.756 s
- 4 MPI, 1 Thread: 6.644, s 7.176, s 7.151 s
- 1 MPI, 4 Threads: 12.496 s, 12.366 s, 12.563 s

Could this be related somehow to #2125?

**Comment by hakonsbm** [comment]
>  Another problem is that the new test by @hakonsbm is not run at all, because it is not yet entered [in the list of Python MPI-tests](https://github.com/hakonsbm/nest-simulator/blob/3dc3d38bb459895b49e44eaf258263d99a100680/testsuite/pytests/test_mpitests.py#L44-L53). I tried to add it but then got some very confusing error output.

The test shouldn't be run with `test_mpitests.py`. The new test,`test_parameters_thread_invariance.py`, works similarly as SLI's `distributed_invariant_assert_or_die` by running `mpitest_test_parameters_thread_invariance.py` with varying number of MPI-processes and checking that the results are invariant. But because the testuite segfaults, the test is not run here.

**Comment by heplesser** [comment]
> > Another problem is that the new test by @hakonsbm is not run at all, because it is not yet entered [in the list of Python MPI-tests](https://github.com/hakonsbm/nest-simulator/blob/3dc3d38bb459895b49e44eaf258263d99a100680/testsuite/pytests/test_mpitests.py#L44-L53). I tried to add it but then got some very confusing error output.
> 
> The test shouldn't be run with `test_mpitests.py`. ... But because the testuite segfaults, the test is not run here.

Right, my mistake. In nosetests' ordering of things your new test comes after the one that crashes and so your test does not show up ... we need to do something with our testsuite ;).

**Comment by stinebuu** [comment]
@hakonsbm @heplesser I have not been following the discussions too closely, is this ready to be reviewed, or should I wait until the problems with catching the segmentation faults in the tests are solved?

**Comment by heplesser** [comment]
@stinebuu You can review, the tests are handled in #2127, which should be ready soon. We can wait with the merge until #2127 is merge in here.

**Review by stinebuu — COMMENTED** [review]
Thanks! I just have a couple of small comments.

File: nestkernel/connection_creator_impl.h
Comment (hakonsbm) [reply nestkernel/connection_creator_impl.h:621]: Some `Parameter::value()` functions need the thread of the target node to know which internal distribution to use for random number generation. This is because distributions have a stored state, so there must be one distribution per thread. See for example https://github.com/nest/nest-simulator/blob/3dc3d38bb459895b49e44eaf258263d99a100680/nestkernel/parameter.cpp#L103-L107

File: testsuite/pytests/mpitest_test_parameters_thread_invariance.py
Comment (hakonsbm) [reply testsuite/pytests/mpitest_test_parameters_thread_invariance.py:32]: Done!

Comment (heplesser) [reply testsuite/pytests/mpitest_test_parameters_thread_invariance.py:32]: @hakonsbm @stinebuu I suggest that we wait until #2127 is merge (should happen any moment) and then adjust this test to the new setup which makes this all a lot simpler. I can help with the test adjustment as soon as the merge sequence #2127 -> master -> this branch is complete.

File: nestkernel/connection_creator_impl.h
Comment (stinebuu) [reply nestkernel/connection_creator_impl.h:621]: Yes, I know, I was just confused about why we only did this for the spatial connections, but I guess we never set weight and delay directly from the builders in `conn_builder` and that is why. Sorry about that :)

**Comment by hakonsbm** [comment]
The failing static code analysis is because of `pynest/examples/brunel_exp_multisynapse_nest.py`. It's not changed in this PR, but because of the merge of master into this branch, it is still checked. The problem is that pycodestyle checks both W503 (line break before binary operator) and W504 (line break after binary operator. It is of course impossible to satisfy both, so the stack check fails.

It seems like ignored warnings should be passed to static_code_analysis.sh as the 17th argument, here
https://github.com/nest/nest-simulator/blob/ae68b6dfa9ebd2c293f15ecb8435f66bb9279bb2/extras/static_code_analysis.sh#L51

But they never are:
https://github.com/nest/nest-simulator/blob/ae68b6dfa9ebd2c293f15ecb8435f66bb9279bb2/extras/ci_build.sh#L126-L129

So, after adding `E402` to the empty ignore list, when running pycodestyle no other warnings are ignored here:
https://github.com/nest/nest-simulator/blob/ae68b6dfa9ebd2c293f15ecb8435f66bb9279bb2/extras/static_code_analysis.sh#L248

File: nestkernel/connection_creator_impl.h
Comment (hakonsbm) [reply nestkernel/connection_creator_impl.h:621]: In `conn_builder` the parameters are wrapped as a `conn_parameter`:

https://github.com/nest/nest-simulator/blob/d3a01157421039113b24759cce81b25a6ff1e532/nestkernel/conn_parameter.cpp#L82-L86

And we send the target node there too:
https://github.com/nest/nest-simulator/blob/d3a01157421039113b24759cce81b25a6ff1e532/nestkernel/conn_builder.cpp#L359-L360

File: nestkernel/parameter.cpp
Comment (heplesser) [inline nestkernel/parameter.cpp:79]: Passing `nullptr` here is not very elegant. Could we make `nullptr` a default value for the argument, so we wouldn't have to pass it here?

**Review by heplesser — CHANGES_REQUESTED** [review]
@hakonsbm One little item in the code, and I sent you a PR replacing the Py- with a SLI-test, since we have no good Python setup for the kind of invariance test we need here.

File: nestkernel/parameter.cpp
Comment (hakonsbm) [reply nestkernel/parameter.cpp:79]: We could, but changing `value( RngPtr, Node* )` to ` value( RngPtr, Node* node = nullptr )` would give a lot of warnings because of unused arguments.

**Review by heplesser — APPROVED** [review]

**Review by stinebuu — APPROVED** [review]

**Pull request merged**
